### ISSUE TITLE
Annotations to mark parameters, returns and more

### ISSUE THREAD

**Issue by Complex2-Liu** [nvarner/typst-lsp#408]
- I have searched the [issues](https://github.com/nvarner/typst-lsp/issues) of this repo and believe that this is not a duplicate.
- I have searched the [discussions](https://github.com/nvarner/typst-lsp/discussions) and believe that my question is not already covered.

## Feature Request

Motivated by package [tidy](https://github.com/Mc-Zen/tidy). See also [LuaLs](https://luals.github.io/wiki/annotations/) for someone familiar with Neovim.

```typst
/// Check if the given value is defined
///
/// @param v any
/// @return boolean
#let has(v) = {
  type(v) != type(none)
}

/// Like html rem
///
/// @param n integer
/// @param fontsize? length The root font size
/// @return length
#let rem(n, fontsize: 11pt) { return n * fontsize }
```

**Comment by astrale-sharp** [comment]
Hey there :)
What's the request here specifically ?

**Comment by Complex2-Liu** [comment]
`rem` is defined as a function, but typst-lsp prompts that it is a constant.

![rem is a function, not constant](https://github.com/nvarner/typst-lsp/assets/40851978/7d8eb8b9-e4fe-4ab1-abc0-b09549f7c05b)

For packages (say `physica`) inside `@preview`, `tensor` is identified correctly as a function.

![packages inside @preview](https://github.com/nvarner/typst-lsp/assets/40851978/44eeeb33-0d0a-44eb-89eb-e2c8a18c2281)

However, when I invoke this function, typst-lsp prompts that it is a constant again.

![tensor is a function, not constant](https://github.com/nvarner/typst-lsp/assets/40851978/40e27378-3a7c-4cd9-a4be-74e03b6f6b27)

The image below shows how LuaLs works. It displays the comment of a function, and the information about parameters of this function: it takes only one parameter, `v`, which can be of any type.

![LuaLs example](https://github.com/nvarner/typst-lsp/assets/40851978/bbe828fd-fb71-430d-8203-c03365ce5226)

**Comment by astrale-sharp** [comment]
I see thanks, it's clearer now ;)

**Comment by astrale-sharp** [comment]
This is probably a good first issue (I haven't investigated but it doesn't seem too complex). What do you think @Myriad-Dreamin ?

**Comment by Myriad-Dreamin** [comment]
typst-lsp just uses [typst-ide](https://github.com/typst/typst/blob/main/crates/typst-ide/src/lib.rs) for autocompletion. So imo, the inperfect is more probably caused by `ypst-ide, and less by typst-lsp. @astrale-sharp Is it time to check it by our awesome ide testing framework, :)?

https://github.com/nvarner/typst-lsp/blob/0d5b9330a1d515fa47456e755f751ea4056282b8/src/server/lsp.rs#L426-L429

**Comment by Myriad-Dreamin** [comment]
Supplement. If I understand webapp's completion correctly, webapp tells me the `add` function is a constant as well (`c` character in the following figure).

![image](https://github.com/nvarner/typst-lsp/assets/35292584/69111990-3732-46e4-b866-2e10818637e4)

**Comment by nvarner** [comment]
Agreed, this is almost certainly a Typst bug. From a quick look at the relevant code, the `scope_completions` function looks a little ad-hoc, and improving its implementation is probably the fix.
https://github.com/typst/typst/blob/main/crates/typst-ide/src/complete.rs#L1280

**Comment by astrale-sharp** [comment]
> @astrale-sharp Is it time to check it by our awesome ide testing framework, :)?

Unfortunately I just added functionality for testing the actual string being autocompleted, not it's kind which is globally a bit more weird to check if there are many possible completions :sweat_smile:

**Comment by astrale-sharp** [comment]
Then I propose we should close this and open a typst issue, any objection @Complex2-Liu ?

**Comment by Enter-tainer** [comment]
We may open a typst issue and keep this open. And this can be closed once the upstream bug get fixed
